controllers/operatorconfig_controller.go

```diff
@@ -54,51 +78,315 @@ type OperatorConfigReconciler struct {
+		OwnerLabel: OwnerConfigName,
+	}
+
+	if err := r.Create(ctx, &obj); err != nil {
+		if !errors.IsAlreadyExists(err) {
+			log.Error(err, fmt.Sprintf("Failed to create %s", obj.Kind), "Namespace", obj.Namespace, "Name", obj.Name)
+			return err
+		}
+	} else {
+		log.Info(fmt.Sprintf("Created %s/%s", obj.Kind, obj.Name))
+	}
+
+	c := pipelinev1beta1.ClusterTask{}
+	if err := r.Get(ctx, types.NamespacedName{Namespace: obj.Namespace, Name: obj.Name}, &c); err != nil {
+		return err
+	}
+
+	if !reflect.DeepEqual(c.Spec, obj.Spec) {
+		c.Spec = obj.Spec
+		if err := r.Update(ctx, &c); err != nil {
+			log.Error(err, fmt.Sprintf("Failed to update %s", obj.Kind), "Namespace", obj.Namespace, "Name", obj.Name)
+			return err
+		}
+	}
+
+	return nil
 }
 
-func configureCranePlugin(ctx context.Context, log logr.Logger) error {
-	return createResourcesFromFile("crane-ui-plugin.yaml", ctx, log)
+func (r *OperatorConfigReconciler) reconcileConsolePlugin(resource *unstructured.Unstructured, ctx context.Context, log logr.Logger, oc *cranev1alpha1.OperatorConfig) error {
+	var obj consolev1alpha1.ConsolePlugin
+	err := runtime.DefaultUnstructuredConverter.
+		FromUnstructured(resource.UnstructuredContent(), &obj)
+	if err != nil {
+		return err
+	}
+
+	obj.Labels = map[string]string{
+		OwnerLabel: OwnerConfigName,
+	}
+
+	if err := r.Create(ctx, &obj); err != nil {
+		if !errors.IsAlreadyExists(err) {
+			log.Error(err, fmt.Sprintf("Failed to create %s", obj.Kind), "Namespace", obj.Namespace, "Name", obj.Name)
+			return err
+		}
+	} else {
+		log.Info(fmt.Sprintf("Created %s/%s", obj.Kind, obj.Name))
+	}
+
+	c := consolev1alpha1.ConsolePlugin{}
+	if err := r.Get(ctx, types.NamespacedName{Namespace: obj.Namespace, Name: obj.Name}, &c); err != nil {
+		return err
+	}
+
+	if !reflect.DeepEqual(c.Spec, obj.Spec) {
+		c.Spec = obj.Spec
+		if err := r.Update(ctx, &c); err != nil {
+			log.Error(err, fmt.Sprintf("Failed to update %s", obj.Kind), "Namespace", obj.Namespace, "Name", obj.Name)
+			return err
+		}
+	}
+
+	return nil
 }
 
-func createClusterTasks(ctx context.Context, log logr.Logger) error {
-	return createResourcesFromFile("manifests.yaml", ctx, log)
+func (r *OperatorConfigReconciler) deleteResourcesFromFile(path string, ctx context.Context) error {
+	var decoder = yaml.NewDecodingSerializer(unstructured.UnstructuredJSONScheme)
+	data, err := getResources(path)
+	if err != nil {
+		return err
+	}
+
+	for _, resource := range data {
+		if len(resource) > 0 {
+			obj := &unstructured.Unstructured{}
+			_, _, err := decoder.Decode([]byte(resource), nil, obj)
+			if err != nil {
+				return err
+			}
+
+			if err = r.Get(ctx, types.NamespacedName{Name: obj.GetName(), Namespace: obj.GetNamespace()}, obj); err != nil {
+				return err
+			}
+
+			if controllerutil.ContainsFinalizer(obj, Finalizer) {
+				controllerutil.RemoveFinalizer(obj, Finalizer)
+				if err = r.Update(ctx, obj); err != nil {
+					return err
+				}
+			}
+
+			err = r.Delete(ctx, obj)
```

Comment:
could it be possible that delete fails here because of the update in 361? The scenario I have in mind is:

1. the object is at resourceVersion: x
2. hits the update at line 361, object goes to resource version: x+1
3. tries to delete at line 366. The obj variable in line 366 contains resource version x, so apiserver rejects it

 I am not sure if this could happen (the unknown for me is does delete request need the object to be at the right resourceVersion), if you could spend a little time inspecting this it would help in understanding this code a little bit more

Reply:
I have not noticed this behavior, as of now cleanup is working so I think resource version should not cause any trouble

Reply:
+1 

For anyone following:

https://github.com/kubernetes-sigs/controller-runtime/blob/6627ea2a1c4a7f6741792438e8333e96364028f1/pkg/client/typed_client.go#L74 object is not used for anything other than metadata.

Additionally, the update function updates the object going into the call with the right information so we should be fine. https://github.com/kubernetes-sigs/controller-runtime/blob/6627ea2a1c4a7f6741792438e8333e96364028f1/pkg/client/typed_client.go#L70